source/extensions/filters/http/admission_control/thread_local_controller.cc

```diff
@@ -0,0 +1,49 @@
+#include "extensions/filters/http/admission_control/thread_local_controller.h"
+
+#include <cstdint>
+
+#include "envoy/common/pure.h"
+#include "envoy/common/time.h"
+#include "envoy/http/codes.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdmissionControl {
+
+static constexpr std::chrono::seconds defaultHistoryGranularity{1};
+
+ThreadLocalControllerImpl::ThreadLocalControllerImpl(TimeSource& time_source,
+                                                     std::chrono::seconds sampling_window)
+    : time_source_(time_source), sampling_window_(sampling_window) {}
+
+void ThreadLocalControllerImpl::maybeUpdateHistoricalData() {
+  // Purge stale samples.
+  while (!historical_data_.empty() && ageOfOldestSample() >= sampling_window_) {
```

Comment:
if you want to same some cycles here i think you can compute the cutoff timestamp and compare each sample with that instead of computing the diff between sampling time and current time for each sample

might not matter, it depends on how many samples are ejected at a time